From 2e10d64764c566393ca0a1045b7ef70d96542252 Mon Sep 17 00:00:00 2001
From: Chris Ward <tjcw@uk.ibm.com>
Date: Tue, 26 Oct 2021 12:04:10 +0100
Subject: [PATCH] treat '-f -' as a request to read commands from stdin in a
 way which works with a pipe or a file redirection

Write to stdout if in a pipe

Need to update the header for vtysh_read_file

Try without a comment
---
 vtysh/vtysh_config.c | 16 +++++++++++-----
 vtysh/vtysh_main.c   |  5 +++--
 2 files changed, 14 insertions(+), 7 deletions(-)

diff --git a/vtysh/vtysh_config.c b/vtysh/vtysh_config.c
index 4b6c648d2..f0416257d 100644
--- a/vtysh/vtysh_config.c
+++ b/vtysh/vtysh_config.c
@@ -513,13 +513,13 @@ void vtysh_config_dump(void)
 }
 
 /* Read up configuration file from file_name. */
-static int vtysh_read_file(FILE *confp)
+static int vtysh_read_file(FILE *confp, int output_fileno)
 {
 	struct vty *vty;
 	int ret;
 
 	vty = vty_new();
-	vty->wfd = STDERR_FILENO;
+	vty->wfd = output_fileno;
 	vty->type = VTY_TERM;
 	vty->node = CONFIG_NODE;
 
@@ -543,7 +543,13 @@ int vtysh_read_config(const char *config_default_dir)
 	FILE *confp = NULL;
 	int ret;
 
-	confp = fopen(config_default_dir, "r");
+	int is_real_file = strcmp(config_default_dir,"-") ;
+	if ( is_real_file == 0 ) {
+		confp = stdin ;
+	}
+	else {
+		confp = fopen(config_default_dir, "r");
+	}
 	if (confp == NULL) {
 		fprintf(stderr,
 			"%% Can't open configuration file %s due to '%s'.\n",
@@ -551,8 +557,8 @@ int vtysh_read_config(const char *config_default_dir)
 		return CMD_ERR_NO_FILE;
 	}
 
-	ret = vtysh_read_file(confp);
-	fclose(confp);
+	ret = vtysh_read_file(confp, is_real_file ? STDERR_FILENO : STDOUT_FILENO );
+	if ( is_real_file ) fclose(confp);
 
 	return (ret);
 }
diff --git a/vtysh/vtysh_main.c b/vtysh/vtysh_main.c
index 1ca219a26..8baa0d8d6 100644
--- a/vtysh/vtysh_main.c
+++ b/vtysh/vtysh_main.c
@@ -559,9 +559,10 @@ int main(int argc, char **argv, char **env)
 	}
 
 	if (inputfile) {
-		vtysh_flock_config(inputfile);
+		int is_real_file = ( strcmp(inputfile, "-") )  ;
+		if (  is_real_file ) vtysh_flock_config(inputfile);
 		ret = vtysh_read_config(inputfile);
-		vtysh_unflock_config();
+		if ( is_real_file ) vtysh_unflock_config();
 		exit(ret);
 	}
 
-- 
2.27.0

